applications/serial_lte_modem/src/mqtt_c/slm_at_mqtt.c

```diff
@@ -217,41 +220,69 @@ static void mqtt_thread_fn(void *arg1, void *arg2, void *arg3)
 	while (true) {
 		if (!ctx.connected) {
 			LOG_WRN("MQTT disconnected");
+			err = 0;
 			break;
 		}
 		err = poll(&fds, 1, mqtt_keepalive_time_left(&client));
 		if (err < 0) {
 			LOG_ERR("ERROR: poll %d", errno);
+			err = -EIO;
 			break;
 		}
-		/* timeout or revent, send KEEPALIVE */
-		(void)mqtt_live(&client);
-
-		if ((fds.revents & POLLIN) == POLLIN) {
-			err = mqtt_input(&client);
-			if (err != 0) {
-				LOG_ERR("ERROR: mqtt_input %d", err);
-				mqtt_abort(&client);
-				break;
-			}
-		}
+		/* MQTT v3.1.1: Note that a Server is permitted to disconnect a Client that it
+		 * determines to be inactive or non-responsive at any time, regardless of the
+		 * Keep Alive value provided by that Client.
+		 */
 		if ((fds.revents & POLLERR) == POLLERR) {
 			LOG_ERR("POLLERR");
-			mqtt_abort(&client);
+			err = -ENETDOWN;
 			break;
 		}
 		if ((fds.revents & POLLHUP) == POLLHUP) {
 			LOG_ERR("POLLHUP");
-			mqtt_abort(&client);
+			err = -ECONNRESET;
 			break;
 		}
 		if ((fds.revents & POLLNVAL) == POLLNVAL) {
 			LOG_ERR("POLLNVAL");
-			mqtt_abort(&client);
+			err = -ENOTCONN;
 			break;
 		}
+		if ((fds.revents & POLLIN) == POLLIN) {
```

Comment:
All of the above errors skip this, which is problematic, because there can also be data left to be read when connection is closed. It might be better to still read the data in mqtt_input, which might even allow for returning of a more exact correct error code, if it sets errno.

Reply:
We can do mqtt_input() before poll abnormal checks. If mqtt_input() returns error we break from there.